### PR TITLE
refactor: DRY up measureText code

### PR DESCRIPTION
**Pull request by markov00**
## Summary

The PR DRYs up the usage of `measureText` and relatives (`TextMeasure` type, `textMeasure` fn etc).

Only one type and one method is now used and available: all the code is available in `packages/charts/src/utils/bbox/canvas_text_bbox_calculator.ts`
The two function to use are:
- `withTextMeasure` (that encapsulate the creation of the `canvas` element and doesn't append the canvas element to the DOM, avoiding the style recalculation and fixing https://github.com/elastic/elastic-charts/issues/1533)
- `measureText` that use an existing canvas (it doesn't save/restore the context so it's only used where we are restoring the context manually)

The removal of the `append`, `remove` calls of the `canvas` elements to compute the text metrics improved the rendering speed in Safari when a high number of data series are used. This, together with the reuse of a single `measureText` function shared on the geometries call reduce the number of DOM Style Recalculation required improving the speed significantly in Safari (Chrome and Firefox gain also a bit of speed). 

The change created a large diff of VRTs (~600) due to the removal of the default 1px padding applied to the text calculation. I have removed that padding because we no longer need to pad the text to "align" the browser rendering: each browser renders slightly differently the text (both on HTML and on canvas) and slight (subpixel/pixel differences are unavoidable). We actually don't care much about these differences because our tests are per-browser.

There are some major (but fine) noticeable differences in the VRTs diffs:
- 1 or 2 pixels shift of the chart caused by the padding removal
- in partition charts, some texts are pushed from within the chart to outside the chart in linked labels (again minor differences)
all-test-ts-baseline-visual-tests-for-all-stories-sunburst-value-formatted-visually-looks-correct-1-diff

- due to the padding removal, some fitted area are shown/hidden (the problem is already known https://github.com/elastic/elastic-charts/issues/1473)

mixed-stories-test-ts-mixed-series-stories-fitting-functions-stacked-charts-area-charts-ordinal-dataset-no-end-value-should-display-correct-fit-for-type-carry-1-diff

fix https://github.com/elastic/elastic-charts/issues/1533

## For reviewers

Please just consider the [TS changes](https://github.com/elastic/elastic-charts/pull/1535/files/8b62993d0fb68bc0f09b1576268b53cc1d8aa6ad), don't waste too much time on VRTs diffs

### CODE REVIEW

**Review by monfera — APPROVED** [review]

File: packages/charts/src/state/selectors/get_legend_size.ts [hunk 1]
```diff
@@ -45,7 +45,12 @@ export const getLegendSizeSelector = createCustomCachedSelector(
     const bbox = withTextMeasure((textMeasure) =>
       labels.reduce(
         (acc, { label, depth }) => {
-          const { width, height } = textMeasure(label, 1, 12, MAGIC_FONT_FAMILY, 1.5, 400);
+          const { width, height } = textMeasure(
+            label,
+            { fontFamily: MAGIC_FONT_FAMILY, fontVariant: 'normal', fontWeight: 400, fontStyle: 'normal' },
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/state/selectors/get_legend_size.ts:50]: probably, we should move it to the Legend theme

File: packages/charts/src/chart_types/heatmap/state/selectors/compute_chart_dimensions.ts [hunk 2]
```diff
@@ -169,17 +150,17 @@ function getTextSizeDimension(
...
       fontWeight: 'bold',
       fontStyle: style.fontStyle ?? 'normal',
-      fontFamily: style.fontFamily,
-      textColor: style.fill,
-      text,
     },
-  ]);
-  return textBox.length === 1 ? textBox[0].width + textPadding : 0;
+    style.fontSize,
+  );
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/chart_types/heatmap/state/selectors/compute_chart_dimensions.ts]: done bcf6879

File: packages/charts/src/chart_types/heatmap/layout/viewmodel/viewmodel.ts [hunk 3]
```diff
@@ -56,12 +56,12 @@ function getValuesInRange(
...
   chartWidth: number,
   formatter: HeatmapSpec['xAxisLabelFormatter'],
-  { padding, fontSize, fontFamily }: HeatmapStyle['xAxisLabel'],
+  style: HeatmapStyle['xAxisLabel'],
 ): number {
   return withTextMeasure((textMeasure) => {
     const labelSample = formatter(Date.now());
-    const { width } = textMeasure(labelSample, horizontalPad(padding), fontSize, fontFamily);
-    const maxTicks = chartWidth / width;
+    const { width } = textMeasure(labelSample, { ...style }, style.fontSize);
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/chart_types/heatmap/layout/viewmodel/viewmodel.ts]: done 897359a

File: packages/charts/src/chart_types/xy_chart/rendering/bars.ts [hunk 4]
```diff
@@ -167,24 +166,28 @@ function computeBoxWidth(
...
+    measureText: TextMeasure;
     width: number;
   },
   displayValueSettings: DisplayValueSpec | undefined,
 ): { fixedFontScale: number; displayValueWidth: number } {
   const fixedFontScale = Math.max(typeof fontSize === 'number' ? fontSize : fontSize.min, 1);
 
-  const computedDisplayValueWidth = textMeasure(text || '', padding, fixedFontScale, fontFamily).width;
+  const computedDisplayValueWidth = measureText(
+    text || '',
```
[the remark this answers is not in the document]
Comment (markov00) [reply packages/charts/src/chart_types/xy_chart/rendering/bars.ts]: done c150965

**Comment by markov00** [comment]
jenkins test this
(let's see if the heatmap change is flaky or not)

**Pull request merged**